rust/lancedb/src/arrow.rs

```diff
@@ -119,3 +128,214 @@ impl<T: arrow_array::RecordBatchReader + Send + 'static> IntoArrow for T {
+    pub fn new(df: DataFrame) -> Self {
+        let fields: Vec<arrow_schema::Field> = df
+            .schema()
+            .into_iter()
+            .map(|(name, dtype)| {
+                arrow_schema::Field::new(
+                    name,
+                    arrow_schema::DataType::from(dtype.to_arrow(false)),
+                    true,
+                )
+            })
+            .collect();
+        // Use pl_flavor = false to use LargeBinary and LargeUtf8 Arrow types instead of
+        // BinaryView and Utf8View types because polars-arrow to arrow-rs conversion
+        // is not yet implemented for BinaryView and Utf8View.
+        // See: https://lists.apache.org/thread/w88tpz76ox8h3rxkjl4so6rg3f1rv7wt for the
+        // differences in the types.
+        PolarsDataFrameRecordBatchReader {
+            chunks: df.iter_chunks(false).collect(),
+            arrow_schema: Arc::new(arrow_schema::Schema::new(fields)),
+            index: 0,
+        }
+    }
+}
+
+#[cfg(feature = "polars")]
+impl Iterator for PolarsDataFrameRecordBatchReader {
+    type Item = std::result::Result<arrow_array::RecordBatch, arrow_schema::ArrowError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index < self.chunks.len() {
+            let columns: Vec<arrow_array::ArrayRef> = self.chunks[self.index]
+                .arrays()
+                .iter()
+                .map(|polars_array| arrow_array::ArrayRef::from(&**polars_array))
+                .collect();
+            self.index += 1;
+            Some(arrow_array::RecordBatch::try_new(
+                self.arrow_schema.clone(),
+                columns,
+            ))
+        } else {
+            None
+        }
+    }
+}
+
+#[cfg(feature = "polars")]
+impl arrow_array::RecordBatchReader for PolarsDataFrameRecordBatchReader {
+    fn schema(&self) -> Arc<arrow_schema::Schema> {
+        self.arrow_schema.clone()
+    }
+
+    fn next_batch(
+        &mut self,
+    ) -> std::prelude::v1::Result<Option<arrow_array::RecordBatch>, arrow_schema::ArrowError> {
+        self.next().transpose()
+    }
+}
+
+/// A trait for converting the result of a LanceDB query into a Polars DataFrame with aligned
+/// chunks. The resulting Polars DataFrame will have aligned chunks, but the series's
+/// chunks are not guaranteed to be contiguous.
+#[cfg(feature = "polars")]
+pub trait IntoPolars {
+    fn into_polars(self) -> impl std::future::Future<Output = Result<DataFrame>> + Send;
+}
+
+#[cfg(feature = "polars")]
+impl IntoPolars for SendableRecordBatchStream {
+    async fn into_polars(self) -> Result<DataFrame> {
+        let arrow_schema = self.schema();
+        let polars_schema = convert_arrow_schema_to_polars_schema(&arrow_schema);
+        let mut acc_df: DataFrame = DataFrame::from(&polars_schema);
+        let record_batches = self.try_collect::<Vec<_>>().await?;
+        for record_batch in record_batches {
```

Comment:
No need to collect if you're just going to iterate.  You should be able to do something like...

```
while let Some(record_batch) = self.next().await? {
  ...
}
```

This still means we iterate the entire stream before returning so its a pretty minor optimization (saves us the one allocation of the vec) but that is inevitable.

Reply:
Done.